File: DataStructure/Tree/q222-count-complete-tree-nodes.py
# 给出一个完全二叉树，求出该树的节点个数。
#
#  说明：
#
#  完全二叉树的定义如下：在完全二叉树中，除了最底层节点可能没填满外，其余每层节点数都达到最大值，并且最下面一层的节点都集中在该层最左边的若干位置。若最底层为
# 第 h 层，则该层包含 1~ 2h 个节点。
#
#  示例:
#
#  输入:
#     1
#    / \
#   2   3
#  / \  /
# 4  5 6
#
# 输出: 6
#  Related Topics 树 二分查找
#  👍 291 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def countNodes(self, root):
        if not root: return 0
        L, R = root.left, root.right
        nL, nR = 1, 1
        while L:
            L = L.left
            nL += 1
        while R:
            R = R.right
            nR += 1  # ↓ (满二叉树)节点数N = 2^h - 1
        if nL == nR:
            return (1 << nL) - 1
        else:
            return self.countNodes(root.left) + 1 + self.countNodes(root.right)
# leetcode submit region end(Prohibit modification and deletion)

File: DataStructure/Tree/q222_count_complete_tree_nodes.java
package DataStructure.Tree;

import java.util.ArrayDeque;
import java.util.Deque;

public class q222_count_complete_tree_nodes {
    // 1. [普通] dfs
    public int countNodes_1(TreeNode root) {
        if (root == null) return 0;
        int cnt_L = countNodes(root.left);
        int cnt_R = countNodes(root.right);
        return 1 + cnt_L + cnt_R;
    }

    // 2. [普通] bfs (层次遍历)
    public int countNodes_2(TreeNode root) {
        if (root == null) return 0;
        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);
        int cnt = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            cnt += size; // 法1 [荐，减少add操作数]
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                // cnt++; // 法2
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
        }
        return cnt;
    }

    // 3.【完全二叉树-性质】 - dfs求子树(until满足满二叉树)高/深h, 节点数=2^h-1
    public int countNodes(TreeNode root) {
        if (root == null) return 0;
        TreeNode left = root.left, right = root.right;
        int depthL = 1, depthR = 1;
        while (left != null) {
            depthL++;
            left = left.left;
        }
        while (right != null) {
            depthR++;
            right = right.right;
        }
        if (depthL == depthR) // 满二叉(子)树【最左孩子深度==最右孩子深度】
            return (2 << (depthL-1)) - 1; // 节点数=2^h-1
        // 否则，下探 直到满二叉子树 // ↓ 不可写left和right（被更新了）
        return 1 + countNodes(root.left) + countNodes(root.right);
    }

    // 4.(变题-only针对满二叉树). 【满二叉树-性质】 - dfs/bfs求树高h, 节点数=2^h-1
    private int maxH = 0;
    public int countNodes_WA(TreeNode root) {
        if (root == null) return 0;
        getHeight(root, 0);
        return 2 << (maxH-1) + 1; // 2^h = Math.pow(2, h) = 2 << (h-1)
        // ↑ 注意: 【2 ^ 2】 = 4 = 2*2 = 【2 << 1】
    }

    private void getHeight(TreeNode root, int height) {
        maxH = height > maxH? height : maxH;
        if (root == null) return ;

        getHeight(root.left, height+1); // 回溯
        getHeight(root.right, height+1);
    }

}
